{% extends "ymap_webtool/base.html" %}

<!--template inheritance: add page title-->
{% block title %}Overview{% endblock %}

<!--template inheritance: in case something else is needed in the head-->
{% block head %}{% endblock %}

<!--template inheritance: add the class "active in the block *_class corresponding to the current page -->
{% block about_class %}active{% endblock %}">

<!--template inheritance: add content -->
{% block content %}

<!--------------------------------------------------------------VERTICAL MENU--------------------------------------------------------------->
           
<div class= "V_menu b_col1">
            <ul class="V_menu_ul b_col1">
                        <li id="menu_overview" class ="V_menu_item active"><a href={% url 'overview' %}>Overview</a></li>
                        <li id="menu_pipeline" class ="V_menu_item "><a href={% url 'pipeline' %}>Pipeline</a></li>
                        <li id="menu_databases" class ="V_menu_item "><a href={% url 'databases' %}>Databases</a></li>
                        <li id="menu_inputformat" class =" V_menu_item"><a href="input_format.html">Input format</a></li>
                        <li id="menu_results" class ="V_menu_item "><a href={% url 'results'%}>Results</a></li>
                        <li id="menu_howtocite" class ="V_menu_item "><a href={% url 'how_to_cite'%}>How to cite</a></li>
            </ul>     
</div>

<!--------------------------------------------------------------CONTENT--------------------------------------------------------------->

<div class= "about_content">
            
YMap facilitates the interpretation of large scale variant data and the correlation of genotype to phenotype.
<br>
It enables the mapping of large sets of mutations to the structural/functional protein features:
<ul>
            <h3>Protein features from <a href="http://www.uniprot.org/">UniProt KB</a> database:</h3>
            <ul>
                        <li>Protein active and binding sites</li>

                        <li>Protein domains</li>
                                        
                        <li>Protein-nucleotide binding domains</li>
                                    
                        <li>Protein structural regions</li>
                            
                         <li>Protein post-translational modifications</li>
            </ul>
            
            <h3>Post translational modifications from <a href="http://ptmfunc.com/">PTMfunc</a>:</h3>
            <ul>
                        <li>Acetylation present on a residue present at the protein interface</li>
                        <li>Acetylation present on a residue important for protein-protein interaction</li>
                        <li>Ubiquitination present on a residue present at the protein interface</li>
                        <li>Ubiquitination present on a residue important for protein-protein interaction</li>
                        <li>Phosphorylation present on a residue present at the protein interface</li>
                        <li>Phosphorylation present on a residue important for protein-protein interaction</li>
            </ul>
            <h3>Post translational modifications from <a href="http://ptmcode.embl.de/">PTMcode 2.0</a>:</h3>
            <ul>
                        <li>PTM located in a regulatory hotspot (as defined by <a href="http://www.cell.com/cell/fulltext/S0092-8674(12)00706-4?_returnURL=http%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0092867412007064%3Fshowall%3Dtrue"> Beltrao et al. 2012</a>) </li>
                        <li>PTM involved in interaction with an other PTM within the same protein</li>
                        <li>PTM involved in nteraction with an other PTM located on an other protein</li>

            </ul> 
</ul>
It also performs a GO enrichment analysis and provides protein's BioGRID ids.
<br>
<br>

</div>
{% endblock %}
