{% extends "ymap_webtool/base.html" %}

<!--template inheritance: add page title-->
{% block title %}Pipeline{% endblock %}

<!--template inheritance: add the class "active in the block *_class corresponding to the current page -->
{% block about_class %}active{% endblock %}">


<!--template inheritance: add content -->
{% block content %}

<!--------------------------------------------------------------VERTICAL MENU--------------------------------------------------------------->
           
<div class= "V_menu b_col1">
            <ul class="V_menu_ul b_col1">
                        <li id="menu_overview" class ="V_menu_item "><a href={% url 'overview' %}>Overview</a></li>
                        <li id="menu_pipeline" class ="V_menu_item active"><a href={% url 'pipeline' %}>Pipeline</a></li>
                        <li id="menu_databases" class ="V_menu_item "><a href={% url 'databases' %}>Databases</a></li>
                        <li id="menu_inputformat" class =" V_menu_item"><a href="input_format.html">Input format</a></li>
                        <li id="menu_results" class ="V_menu_item "><a href={% url 'results'%}>Results</a></li>
                        <li id="menu_howtocite" class ="V_menu_item "><a href={% url 'how_to_cite'%}>How to cite</a></li>
            </ul>     
</div>

<!--------------------------------------------------------------CONTENT--------------------------------------------------------------->

<div class= "about_content">
            
            <h2>
            STEP 1: Mutation at DNA level to protein level (optional)
            </h2>
            <p>
            When the input file contain mutations at DNA level, it is translated to protein level for yMap
            to be able to process. All the non-synonymous mutation are reported in a file that will be use as input by the core
            function of yMap.
            </p>
            
            <h2>
            STEP 2: Mapping the mutations to protein features and post translational modifications.
            </h2>
            <p>
            The mutations at protein level are mapped to protein features and post translational modifications. yMap generates a report file
            for each types of protein features and post translational modifications that contains/overlaps with mutations from the input file.
            
            </p>
             
            <h2>
            STEP 3: GO enrichment analysis
            </h2>
            <p>
            Gene Ontology (GO) enrichment effectively predicts which biological process, cellular component and molecular
            functions are potentially disturbed by a subset of the mutations. It identifies GO terms that are overrepresented in sets of the genes
            to which mutations have been mapped, as compared to the background of the complete yeast genome.
            The significance of enriched terms is computed using the <a href= http://orange-bioinformatics.readthedocs.io/en/latest/reference/go.html>Orange Bioinformatics package</a>. 
            </p>
            
            <h2>
            STEP 4: Biogrid id
            </h2>
            <p>
            The list of BioGrid IDs for each genes to which mutations have been mapped is generated. 
            </p>
            
       
</div>
</div>
{% endblock %}